hpedockerplugin/cmd/cmd_deleteshare.py

```diff
@@ -19,93 +23,107 @@ def __init__(self, file_mgr, share_info):
+                                                    backend_metadata)
+            except Exception as ex:
+                msg = "WARNING: Metadata for backend %s is not " \
+                      "present. Exception: %s" % \
+                      (self._backend, six.text_type(ex))
+                LOG.warning(msg)
+
+    def _fpg_owned_by_docker(self):
+        LOG.info("Checking if FPG %s is owned by Docker..." % self._fpg_name)
+        try:
+            self._fp_etcd.get_fpg_metadata(
+                self._backend, self._cpg_name, self._fpg_name)
+            LOG.info("FPG %s is owned by Docker!" % self._fpg_name)
+            return True
+        except exception.EtcdMetadataNotFound:
+            LOG.info("FPG %s is NOT owned by Docker!" % self._fpg_name)
+            return False
 
     def _remove_quota(self):
         try:
-            share = self._etcd.get_share(self._share_info['name'])
-            if 'quota_id' in share:
-                quota_id = share.pop('quota_id')
+            share_copy = copy.deepcopy(self._share_info)
+            if 'quota_id' in share_copy:
+                quota_id = share_copy.pop('quota_id')
                 self._mediator.remove_quota(quota_id)
-                self._etcd.save_share(share)
+                self._etcd.save_share(share_copy)
         except Exception as ex:
             LOG.error("ERROR: Remove quota failed for %s. %s"
-                      % (self._share_info['name'], six.text_type(ex)))
+                      % (share_copy['name'], six.text_type(ex)))
 
     def _delete_share(self):
         share_name = self._share_info['name']
-        LOG.info("cmd_deleteshare:remove_share: Removing %s..." % share_name)
+        LOG.info("Start delete share %s..." % share_name)
         try:
-            LOG.info("Deleting share %s from backend..." % share_name)
             if self._share_info.get('id'):
+                LOG.info("Deleting share %s from backend..." % share_name)
                 self._mediator.delete_share(self._share_info['id'])
-            LOG.info("Share %s deleted from backend" % share_name)
+                LOG.info("Share %s deleted from backend" % share_name)
+                self._delete_share_from_etcd(share_name)
+        except Exception as e:
+            msg = 'Failed to remove share %(share_name)s from backend: %(e)s'\
+                  % ({'share_name': share_name, 'e': six.text_type(e)})
+            LOG.error(msg)
+
+    def _delete_file_store(self):
+        share_name = self._share_info['name']
+        try:
             LOG.info("Deleting file store %s from backend..." % share_name)
             self._mediator.delete_file_store(self._fpg_name, share_name)
             LOG.info("File store %s deleted from backend" % share_name)
-
         except Exception as e:
-            msg = 'Failed to remove share %(share_name)s from backend: %(e)s'\
+            msg = 'Failed to remove file store %(share_name)s from backend: ' \
```

Comment:
Do we need to raise the exception here?
In such case where share deletion is successful but store deletion fails, shall we consider it as failure of share deletion. Also shall we propagate same status to user/admin so that he/she will be aware of it. He/she can manually go to 3PAR and delete it. 

We need not to worry about etcd record for store as we are not maintaining it.

Reply:
Again, this is a best attempt approach to delete share and all the related entities which cannot happen with raising exception.